### PR TITLE
Add RelationConfig Protocol for use in Relation.create_from

### LINKED ISSUE #9215: [CT-3455] Remove usage of dbt.contracts.graph.nodes from dbt/adapters/base/relation.py

### PR DESCRIPTION
**Pull request by MichelleArk**
resolves #9215

### Solution
* Introduces [RelationConfigProtocol](https://github.com/dbt-labs/dbt-core/pull/9210/files#diff-478bf5088a88fadb1a36b0e846df6ed438c0ca118c1311307cd359d853ba1258R25-R30)
* This protocol is used to consolidate `Relation.create_from`, removing the need to have NodeType-based create_from_node and create_from_source

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

**Comment by MichelleArk** [comment]
After this PR, remaining core imports in dbt/adapters: 

```
❯ python3 scripts/migrate-adapters.py core/dbt/adapters

core/dbt/adapters/protocol.py:
  - dbt.contracts.graph.model_config::BaseConfig
  - dbt.contracts.graph.manifest::Manifest

core/dbt/adapters/factory.py:
  - dbt.mp_context::get_mp_context

core/dbt/adapters/base/connections.py:
  - dbt.contracts.graph.manifest::Manifest

core/dbt/adapters/base/query_headers.py:
  - dbt.context.manifest::generate_query_header_context
  - dbt.contracts.graph.nodes::ResultNode
  - dbt.contracts.graph.manifest::Manifest

core/dbt/adapters/base/impl.py:
  - dbt.contracts.graph.manifest::Manifest
  - dbt.contracts.graph.manifest::MacroManifest
  - dbt.contracts.graph.nodes::ResultNode

migration progress: 95.82% of dbt imports are valid (from adapters or common)
remaining core imports: 10
```

File: core/dbt/adapters/contracts/relation.py [hunk 1]
```diff
@@ -22,6 +22,14 @@ class RelationType(StrEnum):
     Ephemeral = "ephemeral"
 
 
+class RelationConfig(Protocol):
+    name: str
+    database: str
+    schema: str
+    identifier: str
+    quoting_dict: Dict[str, bool]
```
Comment (MichelleArk) [inline core/dbt/adapters/contracts/relation.py:30]: we may want to include meta or tags to accommodate adapter-specific implementations, for example in dbt-duckdb: https://github.com/duckdb/dbt-duckdb/blob/master/dbt/adapters/duckdb/relation.py#L11

But these are the attributes necessary for core <> adapters + postgres. Have carved out an issue to add additional fields later on: https://github.com/dbt-labs/dbt-core/issues/9216

File: core/dbt/adapters/base/impl.py [hunk 2]
```diff
@@ -476,7 +476,7 @@ def _get_catalog_relations(self, manifest: Manifest) -> List[BaseRelation]:
             manifest.sources.values(),
         )
 
-        relations = [self.Relation.create_from(self.config, n) for n in nodes]
+        relations = [self.Relation.create_from(self.config, n) for n in nodes]  # type: ignore[arg-type]
```
Comment (MichelleArk) [inline core/dbt/adapters/base/impl.py:479]: this gets cleaned up in: https://github.com/dbt-labs/dbt-core/pull/9212

File: core/dbt/adapters/base/impl.py [hunk 3]
```diff
@@ -429,7 +429,7 @@ def _get_cache_schemas(self, manifest: Manifest) -> Set[BaseRelation]:
         """
         # the cache only cares about executable nodes
         return {
-            self.Relation.create_from(self.config, node).without_identifier()
+            self.Relation.create_from(self.config, node).without_identifier()  # type: ignore[arg-type]
```
Comment (MichelleArk) [inline core/dbt/adapters/base/impl.py:432]: this gets cleaned up in: https://github.com/dbt-labs/dbt-core/pull/9213

File: core/dbt/adapters/base/relation.py [hunk 4]
```diff
@@ -198,33 +196,14 @@ def quoted(self, identifier):
...
     def add_ephemeral_prefix(name: str):
         return f"__dbt__cte__{name}"
 
     @classmethod
-    def create_ephemeral_from_node(
+    def create_ephemeral_from(
         cls: Type[Self],
-        config: HasQuoting,
-        node: ManifestNode,
+        node: RelationConfig,
```
Comment (colin-rogers-dbt) [inline core/dbt/adapters/base/relation.py:206]:
```suggestion
        relation_config: RelationConfig,
```

File: core/dbt/adapters/base/relation.py [hunk 5]
```diff
@@ -234,47 +213,33 @@ def create_ephemeral_from_node(
...
 
     @classmethod
-    def create_from_node(
+    def create_from(
         cls: Type[Self],
-        config: HasQuoting,
-        node,
-        quote_policy: Optional[Dict[str, bool]] = None,
+        quoting: HasQuoting,
+        config: RelationConfig,
```
Comment (colin-rogers-dbt) [inline core/dbt/adapters/base/relation.py:219]:
```suggestion
        relation_config: RelationConfig,
```

**Review by colin-rogers-dbt — APPROVED** [review]

**Pull request merged**